File: robocode2-server/src/main/java/net/robocode2/mappers/BotsToBotsWithIdMapper.java
package net.robocode2.mappers;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import net.robocode2.schema.events.BotStateWithId;
import net.robocode2.model.Bot;

public final class BotsToBotsWithIdMapper {

	private BotsToBotsWithIdMapper() {}

	public static List<BotStateWithId> map(Set<Bot> bots) {
		List<BotStateWithId> botStates = new ArrayList<>();
		for (Bot bot : bots) {
			botStates.add(BotToBotStateWithIdMapper.map(bot));
		}
		return botStates;
	}
}

File: server/src/main/kotlin/dev/robocode/tankroyale/server/mapper/BotsToBotsWithIdMapper.kt
package dev.robocode.tankroyale.server.mapper

import dev.robocode.tankroyale.schema.BotStateWithId
import dev.robocode.tankroyale.schema.Participant
import dev.robocode.tankroyale.server.model.BotId
import dev.robocode.tankroyale.server.model.IBot

object BotsToBotsWithIdMapper {
    fun map(
        bots: Set<IBot>,
        participantsMap: Map<BotId, Participant>,
        enemyCountMap: Map<BotId, Int /* enemyCount */>,
        debugGraphicsEnableMap: Map<BotId, Boolean /* isDebugGraphicsEnabled */>
    ): List<BotStateWithId> {
        val botStates = mutableListOf<BotStateWithId>()
        bots.forEach { bot ->
            participantsMap[bot.id]?.let { participant ->
                botStates += BotToBotStateWithIdMapper.map(
                    bot,
                    participant.sessionId,
                    enemyCountMap[bot.id] ?: 0,
                    debugGraphicsEnableMap[bot.id] ?: false
                )
            }
        }
        return botStates
    }
}